exercise/exercise_models.py

```diff
@@ -962,9 +967,12 @@ def get_load_url(
             )
         return super().get_load_url(language, request, students, url_name, ordinal)
 
-    def grade(self, request, submission, no_penalties=False, url_name="exercise"):
+    def grade(self, request, submission, uri_prefix=None, no_penalties=False, url_name="exercise"):
         """
         Loads the exercise feedback page.
+        Note: request may be None if grading is initated by async Celery task,
+        i.e., for mass regrade. In such cases uri_prefix is used to compose
+        absolute URI for the grader.
```

Comment:
Now is the time to refactor this grade method so that it can be used without a `request` parameter. I mean the whole `request` parameter is removed. We also have an old issue in building the A+ server URL in the wrong way. We should not trust the request to know the correct name for the server. It is a problem when multiple domains are used (aalto.fi, hut.fi, IP address used directly) and it's always problematic in localhost (Docker internal domain aplus versus the localhost address that is usable in the web browser). In addtition, trusting the request is a security risk since the requests can be forged (source IP address, Host HTTP request header).

The only URL/domain for the server that we can trust is the one we define manually in the server settings (local_settings.py). 

The request variable has been used in many places to build the A+ server's URL, and all of those should be fixed eventually.